x/participationrewards/types/protocol_data.go

```diff
@@ -13,124 +12,47 @@ func NewProtocolData(datatype string, data json.RawMessage) *ProtocolData {
 	return &ProtocolData{Type: datatype, Data: data}
 }
 
+func unmarshalProtocolData(cpd, blank ProtocolDataI, data json.RawMessage) (ProtocolDataI, error) {
```

Comment:
```suggestion
func unmarshalProtocolData(cpd ProtocolDataI, data json.RawMessage) (ProtocolDataI, error) {
        blank := *cpd
	err := json.Unmarshal(data, cpd)
	if err != nil {
		return nil, fmt.Errorf("unable to unmarshal intermediary %s: %w", reflect.TypeOf(cpd).Name(), err)
	}
	if reflect.DeepEqual(cpd, blank) {
		return nil, fmt.Errorf("unable to unmarshal %s from empty JSON object", reflect.TypeOf(cpd).Name())
	}
	return cpd, nil
}
```

would this work, rather than passing in two ProtocolDataI each time?

Reply:
Nice approach though :)

Reply:
Can't do that with ProtocolDataI(*cpd itself won't be accessible) this case would require a cast to its original structural form like cpd.(*ConnectionProtocolData) and then fetch *cpd - it gets kinda ugly that way since we need all the types

One alternative is using generics but then blank has to be slightly hardcoded :- 

`func unmarshalProtocolData[V ProtocolDataI](cpd V, data json.RawMessage) (ProtocolDataI, error) {
	var blank V
	json.Unmarshal([]byte(`{}`), &blank)
	err := json.Unmarshal(data, cpd)
	if err != nil {
		return nil, fmt.Errorf("unable to unmarshal intermediary %s: %w", reflect.TypeOf(cpd).Name(), err)
	}
	if reflect.DeepEqual(cpd, blank) {
		return nil, fmt.Errorf("unable to unmarshal %s from empty JSON object", reflect.TypeOf(cpd).Name())
	}
	return cpd, nil
}`